source/extensions/filters/http/ext_proc/ext_proc.h

```diff
@@ -39,6 +39,32 @@ struct ExtProcFilterStats {
   ALL_EXT_PROC_FILTER_STATS(GENERATE_COUNTER_STRUCT)
 };
 
+inline constexpr absl::string_view ExtProcGrpcStatsName = "ext-proc-filter-state";
+
+class ExtProcGrpcStats : public Envoy::StreamInfo::FilterState::Object {
+public:
+  void record(std::chrono::microseconds latency, ProcessorState::CallbackState callback_state,
+              absl::Status call_status) {
+    switch (callback_state) {
+    case ProcessorState::CallbackState::Idle:
+      header_call_latencies.emplace_back(latency, call_status);
+      break;
+    case ProcessorState::CallbackState::HeadersCallback:
+      header_call_latencies.emplace_back(latency, call_status);
+      break;
+    case ProcessorState::CallbackState::TrailersCallback:
+      trailer_call_latencies.emplace_back(latency, call_status);
+      break;
+    default:
+      body_call_latencies.emplace_back(latency, call_status);
+    }
+  }
+
+  std::vector<std::pair<std::chrono::microseconds, absl::Status>> header_call_latencies;
```

Comment:
There is a single ExtProcGrpcStats instance per stream/FilterState, that ends up being used by both DecodingProcessorState and EncodingProcessorState. AIUI, as currently stands, with the three *_call_latencies fields, this means that request and response latencies will be mixed in the same vector--e.g. header_call_latencies will end up containing entries for both EncodingProcessorState HeadersCallback as well as DecodingProcessorState HeadersCallback. This will make it difficult to separate these measurements for reporting request headers latency vs. response headers latency.

ISTM there are two possible solutions:
A. Use separate FilterState object types for encoding vs. decoding direction.
B. Adjust the fields of ExtProcGrpcStats so that it keeps encoding vs. decoding values separable.

Of these, I think (B) is preferable--a single class is simpler and more idiomatic / in line with how other Filters use FilterState.

If you go with (B), defining 6 separate fields for (header, body, trailer) x (encode, decode) would IMO be a bit clunky. Instead you could define a CallbackType enum { RequestHeaders, RequestBody, RequestTrailers, ResponseHeaders, ResponseBody, ResponseTrailers }, and have a single field like absl::flat_hash_map<std::tuple<CallbackType, absl::Status>, std::vector<std::chrono::microseconds>> call_latencies.

Reply:
Thanks, I will separate the request and response stats.